docs/src/proposals/event_data.md

```diff
@@ -0,0 +1,155 @@
+      ]
+};
+
+var subscription = web3.eth.subscribe('logs', options, function(error, result){
+    if (!error) console.log('got result');
+    else console.log(error);
+}).on("data", function(log){
+    console.log('got data', log);
+}).on("changed", function(log){
+    console.log('changed');
+});
+```
+
+In order to decode the non-indexed fields (the data), the abi of the contract
+is needed. So, the topic is first used to discover what event was used, and
+then the data can be decoded.
+
+### Ethereum Tx in block
+
+The transaction calls events logs. Here is a tx with a single event, with 3
+topics and some data.
+
+```
+{
+  "tx": {
+    "nonce": "0x2",
+    "gasPrice": "0xf224d4a00",
+    "gas": "0xc350",
+    "to": "0x6B175474E89094C44Da98b954EedeAC495271d0F",
+    "value": "0x0",
+    "input": "0xa9059cbb000000000000000000000000a12431d0b9db640034b0cdfceef9cce161e62be40000000000000000000000000000000000000000000000a030dcebbd2f4c0000",
+    "hash": "0x98a67f0a35ebc0ac068acf0885d38419c632ffa4354e96641d6d5103a7681910",
+    "blockNumber": "0xc96431",
+    "from": "0x82f890D638478d211eF2208f3c1466B5Abf83551",
+    "transactionIndex": "0xe1"
+  },
+  "receipt": {
+    "gasUsed": "0x74d2",
+    "status": "0x1",
+    "logs": [
+      {
+        "address": "0x6B175474E89094C44Da98b954EedeAC495271d0F",
+        "topics": [
+          "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
+          "0x00000000000000000000000082f890d638478d211ef2208f3c1466b5abf83551",
+          "0x000000000000000000000000a12431d0b9db640034b0cdfceef9cce161e62be4"
+        ],
+        "data": "0x0000000000000000000000000000000000000000000000a030dcebbd2f4c0000"
+      }
+    ]
+  }
+}
+```
+
+### Further considerations
+
+In Ethereum, events are stored in blocks. Events mark certain state changes in
+smart contracts. This serves two purposes:
+
+ - Listen to events (i.e. state changes) as they happen by reading new blocks
+   as they are published
+ - Re-read historical events by reading old blocks
+
+So far example, smart contracts may emit changes as they happen but never the
+complete state, so the only way to recover the entire state of the contract
+is by re-reading all events from the chain. So an application will read events
+from block 1 or whatever block the application was deployed at and then use
+that state for local processing. This is a local cache and may re-populated
+from the chain at any point.
+
+## Proposed Solution
+
+Events should be added to the program log. The program log should include:
+
+ - program_id that generate the data
+ - base64 encoded data (more than one permitted)
+
+So if we encoding the topics first, followed by the data then the event in the
+tx above would look like:
+```
+program event: GaBtBJvmS4Arjj5W1NmFcyvPjsHN38UGYDq2MDwbs9Qu 3fJSrRviyJtpwrBo/DeNqpUrpFjxKEWKPVaTfUjs8AAAAAAAAAAAAAAACC+JDWOEeNIR7yII88FGa1q/g1UQAAAAAAAAAAAAAAAKEkMdC522QANLDN/O75zOFh5ivk AAAAAAAAAAAAAAAAAAAAAAAAAAAAAACgMNzrvS9MAAA=
+```
```

Comment:
To be clear, are all of the topics encoded as one continuous piece of data?  And is the idea that an event listener would slice this up into 32-byte chunks and then check against each one of those?  I wonder if some more structure might make it clearer, something like: 

```
program event: GaBtBJvmS4Arjj5W1NmFcyvPjsHN38UGYDq2MDwbs9Qu topics 3fJSrRviyJtpwrBo/DeNqpUrpFjxKEWKPVaTfUjs8 AAAAAAAAAAAAAAACC+JDWOEeNIR7yII88FGa1q/g1UQ AAAAAAAAAAAAAAAKEkMdC522QANLDN/O75zOFh5ivk data AAAAAAAAAAAAAAAAAAAAAAAAAAAAAACgMNzrvS9MAAA=
```

(I hope I got the spaces right in the topics part) Would that potentially be clearer for a solidity dev, or would they not care?  On the other hand, this approach would make it look a lot more like eth, which is probably what you were trying to avoid :smile:

Reply:
I'm am concerned about making it too eth-centric.

Maybe we could do something like this:

```
struct event_field {
    const uint8_t *data;
    uint64_t data_length;
    const char *tag;
    uint64_t tag_length;
};
void sol_event(struct event_fields fields, uint64_t length);
```

For Solidity events, the BPF could add the "topics" add "data" tag. I think should be printed with a colon so it is clear that what the tag refers to:

```
program event: GaBtBJvmS4Arjj5W1NmFcyvPjsHN38UGYDq2MDwbs9Qu topics:3fJSrRviyJtpwrBo/DeNqpUrpFjxKEWKPVaTfUjs8AAAAAAAAAAAAAAACC+JDWOEeNIR7yII88FGa1q/g1UQAAAAAAAAAAAAAAAKEkMdC522QANLDN/O75zOFh5ivk data:AAAAAAAAAAAAAAAAAAAAAAAAAAAAAACgMNzrvS9MAAA=
```

Reply:
As long as the event is formatted consistently I don't think folks need the tags, and "topics:" terminology is eth-centric

Reply:
The idea is that that tags would be provided by the BPF program, not hardcoded in solana. I'm undecided though, it seems a bit complex.